Source/DafnyLanguageServer/Workspace/ChangeProcessors/Relocator.cs

```diff
@@ -4,47 +4,86 @@
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 
 namespace Microsoft.Dafny.LanguageServer.Workspace.ChangeProcessors {
-  public class SymbolTableRelocator : ISymbolTableRelocator {
+  public class Relocator : IRelocator {
     private readonly ILogger logger;
     private readonly ILogger<SymbolTable> loggerSymbolTable;
 
-    public SymbolTableRelocator(
-      ILogger<SymbolTableRelocator> logger,
+    public Relocator(
+      ILogger<Relocator> logger,
       ILogger<SymbolTable> loggerSymbolTable
       ) {
       this.logger = logger;
       this.loggerSymbolTable = loggerSymbolTable;
     }
 
-    public SymbolTable Relocate(SymbolTable originalSymbolTable, DidChangeTextDocumentParams changes, CancellationToken cancellationToken) {
-      return new ChangeProcessor(logger, loggerSymbolTable, originalSymbolTable, changes.ContentChanges, cancellationToken).MigrateSymbolTable();
+    public SymbolTable RelocateSymbols(SymbolTable originalSymbolTable, DidChangeTextDocumentParams changes, CancellationToken cancellationToken) {
+      return new ChangeProcessor(logger, loggerSymbolTable, changes.ContentChanges, cancellationToken).MigrateSymbolTable(originalSymbolTable);
+    }
+
+    public IReadOnlyList<Diagnostic> RelocateDiagnostics(IReadOnlyList<Diagnostic> originalDiagnostics, DidChangeTextDocumentParams changes, CancellationToken cancellationToken) {
+      return new ChangeProcessor(logger, loggerSymbolTable, changes.ContentChanges, cancellationToken).MigrateDiagnostics(originalDiagnostics);
     }
 
     private class ChangeProcessor {
       private readonly ILogger logger;
-      private readonly SymbolTable originalSymbolTable;
       private readonly Container<TextDocumentContentChangeEvent> contentChanges;
       private readonly CancellationToken cancellationToken;
       private readonly ILogger<SymbolTable> loggerSymbolTable;
 
       public ChangeProcessor(
         ILogger logger,
         ILogger<SymbolTable> loggerSymbolTable,
-        SymbolTable originalSymbolTable,
         Container<TextDocumentContentChangeEvent> contentChanges,
         CancellationToken cancellationToken
       ) {
         this.logger = logger;
-        this.originalSymbolTable = originalSymbolTable;
         this.contentChanges = contentChanges;
         this.cancellationToken = cancellationToken;
         this.loggerSymbolTable = loggerSymbolTable;
       }
 
-      public SymbolTable MigrateSymbolTable() {
+      public IReadOnlyList<Diagnostic> MigrateDiagnostics(IReadOnlyList<Diagnostic> originalDiagnostics) {
+        return contentChanges.Aggregate(originalDiagnostics, MigrateDiagnostics);
+      }
+
+      private IReadOnlyList<Diagnostic> MigrateDiagnostics(IReadOnlyList<Diagnostic> originalDiagnostics, TextDocumentContentChangeEvent change) {
+        var result = new List<Diagnostic>();
+        foreach (var diagnostic in originalDiagnostics) {
+          cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I suggest extracting the body of this foreach-loop. Besides more clarity, it allows omitting the manual filling of the `result` list since it can be replaced with a LINQ statement.